### PR TITLE
avoid running setuidgid when not root

### LINKED ISSUE #327: cannot run as non-root
```
herokuish procfile start web
setuidgid: fatal: unable to setgid: permission denied
app@48c640052722:/$ id
uid=1000(app) gid=1000(app) groups=1000(app)

cat Procfile 
web: java $JAVA_OPTIONS -jar start.jar
```

Also https://cr.yp.to/daemontools/setuidgid.html states that setuidgid can be run by root only, hence it seems herokuish can only be run as root? Should it not be possible to run as non-root, and specifically skip the setuidgid step if already being the correct user?

OpenShift typically requires containers to run as non-root, hence the request for the feature.

### PR DESCRIPTION
**Pull request by davidkarlsen**
Fixes #327

---
This change is [Reviewable](https://reviewable.io/reviews/gliderlabs/herokuish/328)

### CODE REVIEW

**Comment by michaelshobbs** [comment]
@davidkarlsen how are you executing herokuish? currently there is no expectation to run herokuish as a non-root user.

side note: this change would mean that child processes started would not receive signals from docker commands if the container was not started as root.

cc:  @josegonzalez

**Comment by davidkarlsen** [comment]
I have a baseimages java->jetty, where the jetty image has a CMD like this:
```
cat app/run 
#!/bin/bash

test -z $APP && echo "Please define the APP variable to identify application" && exit 1

if [[ -z $CONSUL_ARGS ]]; then
  exec herokuish procfile start web
else
  exec envconsul -sanitize -consul consul:8500 $CONSUL_ARGS herokuish procfile start web
fi
```

and a Procfile of:
``` 
web: java $JAVA_OPTIONS -jar start.jar
```

so that consumers of the baseimage can just drop in their .war and their config.
What is really neat about herokuish is that it will source the .profile.d dir as well, which has some sensible defaults, as well as can source application specific vars (which are not environment-dependant / no need to -e inject them via docker).

**Comment by michaelshobbs** [comment]
> Why would the child not receive signals? It's still exec'ed: exec - just not exec setuidgid (as this requires root).

oh geez i'm an idiot!! this is why i shouldn't code review at the same time i'm trying to work on a big project. apologies for the confusion here. 😞

**Comment by michaelshobbs** [comment]
More specifically though, how do you herokuish as a non-root user? the point of `exec setuidgid "$unprivileged_user" $(eval echo "$@")` is to run your app as a non-root user.  `herokuish` even does it by default as the `herokuishuser`. You can also customize what user you want to run your app as by defining the `$USER` env variable on container start. 

Seems like you could create an app user in your base image and on execution of the `herokuish` container set `$USER` to that user. Thoughts?

**Comment by davidkarlsen** [comment]
See explanation above. The container is already running as a non-root user, which is a openshift "requirement" (not strictly required, but recommended) - herokuish is used for:
* procfile functionality
* parsing and setting of envs through profile

**Comment by davidkarlsen** [comment]
Also see https://github.com/davidkarlsen/herokuish/tree/withtest - which is the closest we get to a workable unit-test - since you required one - although have no existing test for this function...

**Comment by davidkarlsen** [comment]
And yes, I use the USER directive. But the thing to notice here is to _avoid the setuidgid_ if already a non-root user (through USER directive, or -u when running container)

**Comment by michaelshobbs** [comment]
> ... since you required one - although have no existing test for this function...

as a maintainer of open source software it is my job to help push the project forward either by personal code contribution or by asking contributors to do so.

**Comment by michaelshobbs** [comment]
Regarding this particular PR, it honestly seems out of scope for the project to me but I think I could be convinced otherwise. @josegonzalez what say you?

**Comment by josegonzalez** [comment]
I'd rather we _not_ do automatic detection of non-root user. I'm no linux pro, but I could _maybe_ imagine a case where the user in use (heh) had root-like privileges, so we'd end up _not_ actually having a secure system. I can also envision a user doing something really cockamamie with Dokku and having a bad time.

I agree with @michaelshobbs re: requiring a test. The old behavior is implicitly tested - we have thousands of Dokku users with herokuish _as is_ in production. New behavior should always be accompanied by a test - both to prove that existing setups continue working as expected and so that any new behavior is not removed by a future refactor. I think thats a pretty fair thing to expect.

> Writing tests is actually why I'm not likely to work on some of the open issues; if I'm not being paid to work and the project already handles my needs, I'm totally not going out of my way to work on features that need tests :D

Maybe a good way forward would be for us to check for `[[ "$HEROKUISH_ SETUIDGUID" == "false" ]]`? You're then *explicitly* opting out, and the test could set the env var and verify that the `setuidgid` call was not invoked and ditto with it not set. Since this isn't the 80% case (or even 90%) of our userbase, I think that would be a fair compromise.

**Comment by davidkarlsen** [comment]
I think it is important to understand setuidgid:
http://manpages.ubuntu.com/manpages/trusty/man8/setuidgid.8.html
https://cr.yp.to/daemontools/setuidgid.html
it can be run by root - only. There is no "root-like".

All I am saying is that the function in question (procfile exec) is as of today - before my requested change - without a test - you are free to use mine either you pull the change or not - there is no other way to test this (as the code is today) because you are calling exec - which switches the current process into the child.

Anyway - I think the HEROKUISH_ SETUIDGUID is a nice explicit way to solve it - so I go down that path and you are OK with merging it?

**Comment by josegonzalez** [comment]
Yep that change seems reasonable to me.

File: include/procfile.bash [hunk 1]
```diff
@@ -52,7 +52,11 @@ procfile-exec() {
 	cd "$app_path" || return 1
 	# unprivileged_user is defined in outer scope
 	# shellcheck disable=SC2154,SC2046
-	exec setuidgid "$unprivileged_user" $(eval echo "$@")
+	[[ "$HEROKUISH_SETUIDGUID" == "false" ]]; then
```
[the remark this answers is not in the document]
Comment (josegonzalez) [reply include/procfile.bash]: Needs an `if` statement.

**Comment by josegonzalez** [comment]
Now just needs:

- a test as we described above
- some docs in our readme

And we're good to go :)

**Comment by davidkarlsen** [comment]
See https://github.com/gliderlabs/herokuish/pull/329 - thanks for coop :)

**Comment by davidkarlsen** [comment]
I can merge that into this one if you want to preserve history in one place.

**Comment by josegonzalez** [comment]
Yeah lets have everything go into this PR :)

**Comment by davidkarlsen** [comment]
@josegonzalez done

**Comment by michaelshobbs** [comment]
Released in v0.3.35. Thanks for seeing this PR through!

**Pull request merged**